File: Linked List/83. Remove Duplicates from Sorted List.py
"""
83. Remove Duplicates from Sorted List
Linked List

Description:
Given a sorted linked list, delete all duplicates such that each element appear only once.

Example 1:
Input: 1->1->2
Output: 1->2

Example 2:
Input: 1->1->2->3->3
Output: 1->2->3

Similar Questions:
Remove Duplicates from Sorted List II - Medium
"""

#Solution:
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        
        """
        Method: Brute force
                Create two pointers.
                We move the first pointer in the main loop.
                If the second pointer points to a duplicate, we then skip that node and keep moving it until we reach to a new node.
                Then we link the first pointer to the second by cur.next = cur2.
        """
        if head is None or head.next is None:
            return head
        
        cur = head
        while cur:
            cur2 = cur

            while cur2 and cur2.val == cur.val:
                cur2 = cur2.next
            cur.next = cur2
            cur = cur.next
        return head

File: Linked List/83. Remove Duplicates from Sorted List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// Method 1: using a set to save distinct values.
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if(head==NULL || head->next==NULL){
            return head;
        }
        set<int> s;
        ListNode *cur = head;
        while(cur!=NULL){
            s.insert(cur->val);
            while(cur->next!=NULL && s.find(cur->next->val)!=s.end()){
                cur->next = cur->next->next;
            }
            cur = cur->next;
        }
        return head;
    }
};


// Method 2: Save all the values to a set, and transfer those distinct values to a new list.
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if(head==NULL || head->next==NULL){
            return head;
        }
        ListNode *head2 = new ListNode(0);
        ListNode *cur = head2;
        set<int> s;
        while(head!=NULL){
            s.insert(head->val);
            head = head->next;
        }
        for(auto & val:s){
            ListNode *newNode = new ListNode(val);
            cur->next = newNode;
            cur = cur->next;
        }
        return head2->next;
    }
};
